File: p463e/island_perimeter.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var islandPerimeter = function (grid) {
  const [rows, cols] = [grid.length, grid[0].length];
  let ans = 0;
  for (let i = 0; i < rows; ++i) {
    for (let j = 0; j < cols; ++j) {
      if (grid[i][j] === 1) {
        ans += 4;
        if (i < rows - 1 && grid[i + 1][j] === 1) ans -= 2;
        if (j < cols - 1 && grid[i][j + 1] === 1) ans -= 2;
      }
    }
  }
  return ans;
};
// TEST
[
  [[1]],

  [
    [0, 0],
    [1, 1],
  ],

  [
    [0, 1],
    [1, 1],
  ],

  [
    [0, 1, 0, 0],
    [1, 1, 1, 0],
    [0, 1, 0, 0],
    [1, 1, 0, 0],
  ],
].forEach((t) => {
  console.log('Perimeter of the island ->', islandPerimeter(t));
});

File: p463e/island_perimeter.py
from typing import List


class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        ans, R, C = 0, len(grid), len(grid[0])
        for i in range(R):
            for j in range(C):
                if grid[i][j] == 1:
                    p = 4
                    if i > 0 and grid[i - 1][j] == 1:
                        p -= 2
                    if j > 0 and grid[i][j - 1] == 1:
                        p -= 2
                    ans += p
        return ans


# TESTS
tests = [
    ([[0, 1, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], 4),
    ([[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], 6),
    ([[0, 1, 0, 0], [1, 1, 1, 0], [0, 1, 0, 0], [1, 1, 0, 0]], 16),
]
for t in tests:
    sol = Solution()
    actual = sol.islandPerimeter(t[0])
    print("Perimeter of the island ->", actual)
    assert actual == t[1]
